apps/src/code-studio/initApp/loadApp.js

```diff
@@ -276,13 +276,51 @@ function tryToUploadShareImageToS3({image, level}) {
 }
 
 /**
+ * Loads project and checks to see if it is abusive.
+ * @returns {Promise.<AppOptionsConfig>} Resolves when project has loaded and is
+ * not abusive. Never resolves if abusive.
+ */
+function loadProjectAndCheckAbuse(appOptions) {
+  return new Promise((resolve, reject) => {
+    project.load().then(() => {
+      if (project.hideBecauseAbusive() && !appOptions.canResetAbuse) {
+        renderAbusive(window.dashboard.i18n.t('project.abuse.tos'));
+        return;
+      }
+      if (project.hideBecausePrivacyViolationOrProfane()) {
+        renderAbusive(window.dashboard.i18n.t('project.abuse.policy_violation'));
+        return;
+      }
+      resolve(appOptions);
+    });
+  });
+}
```

Comment:
This could probably be done more cleanly if project.load returned an ES6 Promise instead of a $.Deferred (which is a similar but slightly different concept). It's also a little weird in that we depend on it not resolving/rejecting in the case where we find abuse. This particular weirdness also existed before, it just wasnt quite as evident it was there.

Reply:
Agreed, though what you have here is fine too. I was going to suggest a way to wrap it here, but it looks like it would be pretty straight forward to convert from Deferred to Promise.